server/src/lib/model/achievements.ts

```diff
@@ -0,0 +1,128 @@
+  invite_signup: `
+    SELECT (invitee.id IS NOT NULL) AS win_bonus, invitor.client_id AS bonus_winner
+    FROM challenges
+    LEFT JOIN enroll invitee ON invitee.client_id = ?
+        AND invitee.invited_by = ? AND invitee.enrolled_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN enroll invitor ON invitor.url_token = ?
+    WHERE challenges.url_token = ?
+    `,
+  // Arguments: [client_id, client_id, challenge]
+  // To earn an three_day_streak bonus, user has to satisfy all the following conditions:
+  // 1) has been enrolled;
+  // 2) not earned three_day_streak bonus before;
+  // 3) already set a customer goal
+  // 4) already reached a 3-day-streak goal OR haven't reached such a goal but has streak activity that lasts for AT LEAST 3 days
+  //    just in case no such bonus are missed
+  three_day_streak: `
+    SELECT
+        (enroll.client_id IS NOT NULL) AND (earn.client_id IS NULL) AND (custom_goals.id IS NOT NULL) AND ((reached_goals.client_id IS NOT NULL) OR (streaks.id IS NOT NULL)) AS win_bonus,
+        ? AS bonus_winner
+    FROM challenges
+    LEFT JOIN achievements ON achievements.name = 'three_day_streak' AND challenges.id = achievements.challenge_id
+    LEFT JOIN enroll ON enroll.client_id = ?
+        AND challenges.id = enroll.challenge_id
+        AND enroll.enrolled_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN earn ON enroll.client_id = earn.client_id
+        AND earn.achievement_id = achievements.id
+        AND earned_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN custom_goals ON custom_goals.client_id = enroll.client_id
+        AND custom_goals.created_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN reached_goals ON reached_goals.client_id = enroll.client_id
+        AND reached_goals.type = 'streak'
+        AND reached_goals.count >= 3
+        AND reached_goals.reached_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN streaks ON streaks.client_id = enroll.client_id
+        AND streaks.started_at BETWEEN start_date AND TIMESTAMPADD(DAY, 18, start_date)
+        AND TIMESTAMPDIFF(DAY, streaks.started_at, streaks.last_activity_at) > 3
+    WHERE challenges.url_token = ?
+  `,
+  // Arguments: [challenge, client_id]
+  // To earn an first_contribution, user's contribution must be exact 1.
+  first_contribution: `
+    SELECT count(distinct clips.id) + count(distinct votes.id) = 1 AS win_bonus, user_clients.client_id AS bonus_winner
+    FROM user_clients
+    LEFT JOIN challenges ON url_token = ?
+    LEFT JOIN clips ON clips.client_id = user_clients.client_id AND clips.created_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN votes ON votes.client_id = user_clients.client_id AND votes.created_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    WHERE user_clients.client_id = ?
+    GROUP BY user_clients.client_id
+    `,
```

Comment:
I think we should make this available to preexisting users as well, by awarding the achievement to anyone who:

- Contributes, and,
- doesn't have the achievement yet.

It shouldn't need to be their first contribution _ever,_ just the first contribution within the challenge.

Reply:
Err, I use LEFT JOIN clips and votes with ON conditions.

Check out this part "`clips.created_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)`", this will select the clips that are created during the challenge. So we don't count clips that are created before and after the challenge.
Same as the votes.

Is this the suits the requirement?